module/icp/algs/impl/impl.c

```diff
@@ -0,0 +1,175 @@
+	size_t i, c;
+	uint32_t max_priority = 0;
+
+	/* Copy from available to supported */
+	for (i = 0, c = 0; i < conf->available_n; i++) {
+		curr_impl = conf->available[i];
+
+		if (curr_impl->is_supported()) {
+			conf->supported[c++] = curr_impl;
+
+			/* Keep track of fastest */
+			if (curr_impl->priority > max_priority) {
+				max_priority = curr_impl->priority;
+				fast_impl = curr_impl;
+			}
+		}
+	}
+	conf->supported_n = c;
+
+	/* Copy the fastest implementation. */
+	ASSERT3P(fast_impl, !=, NULL);
+	memcpy(&conf->fastest, fast_impl, sizeof (*fast_impl));
+	strlcpy(conf->fastest.name, "fastest", ALG_IMPL_NAME_MAX);
+
+	/*
+	 * Finish initialization. At this, user_sel_impl can only be one of
+	 * alg_impl_opts, because initialized is not yet set, so
+	 * alg_impl_set would have returned EINVAL for any other value.
+	 */
+	conf->cycle_impl_idx = 0;
+	atomic_swap_32(&conf->icp_alg_impl, conf->user_sel_impl);
+	conf->initialized = B_TRUE;
+}
+
+int
+alg_impl_get(alg_impl_conf_t *conf, char *buffer)
+{
+	int cnt = 0;
+	size_t i = 0;
+	char *fmt = NULL;
+	const uint32_t impl = ALG_IMPL_READ(conf->icp_alg_impl);
+
+	ASSERT(conf->initialized);
+
+	/* First check mandatory options. */
+	for (i = 0; i < ARRAY_SIZE(alg_impl_opts); i++) {
+		fmt = (impl == alg_impl_opts[i].sel) ? "[%s] " : "%s ";
+		cnt += sprintf(buffer + cnt, fmt, alg_impl_opts[i].name);
+	}
+
+	/* Then check all supported. */
+	for (i = 0; i < conf->supported_n; i++) {
+		fmt = (i == impl) ? "[%s] " : "%s ";
+		cnt += sprintf(buffer + cnt, fmt, conf->supported[i]->name);
+	}
+
+	return (cnt);
+}
+
+int
+alg_impl_set(alg_impl_conf_t *conf, const char *val)
+{
+	int err = -EINVAL;
+	char req_name[ALG_IMPL_NAME_MAX];
+	uint32_t impl = ALG_IMPL_READ(conf->user_sel_impl);
+	size_t i;
+
+	/* Sanitize input. */
+	i = strnlen(val, ALG_IMPL_NAME_MAX);
+	if (i == 0 || i >= ALG_IMPL_NAME_MAX) {
+		return (err);
+	}
+
+	/* Copy and NULL terminate. */
+	strlcpy(req_name, val, ALG_IMPL_NAME_MAX);
+	while (i > 0 && isspace(req_name[i-1])) {
+		i--;
+	}
+	req_name[i] = '\0';
+
+	/* First check mandatory options. */
+	for (i = 0; i < ARRAY_SIZE(alg_impl_opts); i++) {
+		if (strcmp(req_name, alg_impl_opts[i].name) == 0) {
+			impl = alg_impl_opts[i].sel;
+			err = 0;
+			break;
+		}
+	}
+
+	/* Then check all supported impl if init() was already called. */
+	if (err != 0 && conf->initialized) {
+		for (i = 0; i < conf->supported_n; i++) {
+			if (strcmp(req_name, conf->supported[i]->name) == 0) {
+				impl = i;
+				err = 0;
+				break;
+			}
+		}
+	}
+
+	/* If success, then set depending on configuration state. */
+	if (err == 0) {
+		if (conf->initialized) {
+			atomic_swap_32(&conf->icp_alg_impl, impl);
+		} else {
+			atomic_swap_32(&conf->user_sel_impl, impl);
+		}
+	}
+
+	return (err);
+}
+
+const alg_impl_ops_t *
+alg_impl_get_ops(alg_impl_conf_t *conf)
+{
+	const alg_impl_ops_t *ops = conf->generic;
+
+	/* Check if we are not allowed to use faster versions. */
+	if (!kfpu_allowed()) {
```

Comment:
When is kfpu now allowed?

Will this still be able to use the faster versions most of the time?

Reply:
Support will normally be available, but we need to be able to fallback to the generic implementation for less common architectures (e.g. sparc64).